peekingduck/pipeline/nodes/model/jde_mot/jde_files/utils/utils.py

```diff
@@ -0,0 +1,256 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Original copyright (c) 2019 ZhongdaoWang
+
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+"""
+Utilities functions
+"""
+
+from typing import Any, List, Tuple, Union
+import cv2
+import numpy as np
+import torch
+from torchvision.ops import nms
+
+# pylint: disable=invalid-name, no-member
+
+
+def letterbox(
+    img: np.ndarray,
+    height: int = 608,
+    width: int = 1088,
+    color: Tuple[float, float, float] = (127.5, 127.5, 127.5),
+) -> Tuple[np.ndarray, float, float, float]:
+    """Resize a rectangular image to a padded rectangular.
+
+    Args:
+        img (np.ndarray): Image frame
+        height (int, optional): Height of padded image. Defaults to 608.
+        width (int, optional): Width of padded image. Defaults to 1088.
+        color (Tuple[float, float, float], optional): Colour of border
+            around image. Defaults to (127.5, 127.5, 127.5).
+
+    Returns:
+        Tuple[np.ndarray, float, float, float]: Padded rectangular image.
+    """
+    shape = img.shape[:2]  # shape = [height, width]
+    ratio = min(float(height) / shape[0], float(width) / shape[1])
+    new_shape = (
+        round(shape[1] * ratio),
+        round(shape[0] * ratio),
+    )  # new_shape = [width, height]
+    dw = (width - new_shape[0]) / 2  # width padding
+    dh = (height - new_shape[1]) / 2  # height padding
+    top, bottom = round(dh - 0.1), round(dh + 0.1)
+    left, right = round(dw - 0.1), round(dw + 0.1)
+    img = cv2.resize(img, new_shape, interpolation=cv2.INTER_AREA)  # resized, no border
+    img = cv2.copyMakeBorder(
+        img, top, bottom, left, right, cv2.BORDER_CONSTANT, value=color
+    )  # padded rectangular
+    return img, ratio, dw, dh
+
+
+def scale_coords(
+    img_size: List[Any], coords: torch.Tensor, img0_shape: Tuple[Any, Any]
```

Comment:
more detailed types of element of List and Tuple

Reply:
After checking by printing type of element in `List` and `Tuple`, changed `Any` to `int`